src/shared/Microsoft.Git.CredentialManager/Interop/LibGit2.cs

```diff
@@ -0,0 +1,154 @@
+                    case GIT_ENOTFOUND: // If the file was not found we should just continue
+                        break;
+                    default:
+                        git_config_free(config);
+                        ThrowIfError(error, nameof(git_config_add_file_ondisk));
+                        return null;
+                }
+            }
+
+            return new LibGit2Configuration(config, repositoryPath);
+        }
+
+        public string GetRepositoryPath(string path)
+        {
+            var buf = new git_buf();
+            int error = git_repository_discover(buf, path, true, null);
+
+            try
+            {
+                switch (error)
+                {
+                    case GIT_OK:
+                        return buf.ToString();
+                    case GIT_ENOTFOUND:
+                        return null;
+                    default:
+                        ThrowIfError(error, nameof(git_repository_discover));
+                        return null;
+                }
+            }
+            finally
+            {
+                git_buf_dispose(buf);
+            }
+        }
+
+        private void Dispose(bool disposing)
+        {
+            git_libgit2_shutdown();
+        }
+
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        ~LibGit2()
+        {
+            Dispose(false);
+        }
+    }
+
+    internal class LibGit2Configuration : IGitConfiguration
+    {
+        private readonly unsafe git_config* _config;
+        private readonly unsafe git_config* _snapshot;
+
+        internal unsafe LibGit2Configuration(git_config* config, string repositoryPath)
+        {
+            _config = config;
+            RepositoryPath = repositoryPath;
+
+            // Create snapshot for reading values
+            _snapshot = CreateSnapshot(_config);
+        }
+
+        public string RepositoryPath { get; }
+
+        public string this[string name] => GetString(name);
+
+        public unsafe string GetString(string name)
+        {
+            int result = git_config_get_string(out string value, _snapshot, name);
+
+            switch (result)
+            {
+                case GIT_ENOTFOUND:
+                    return null;
+                default:
+                    ThrowIfError(result, nameof(git_config_get_string));
+                    break;
+            }
+
+            return value;
+        }
+
+        private void Dispose(bool disposing)
+        {
+            unsafe
+            {
+                git_config_free(_snapshot);
+                git_config_free(_config);
+            }
+        }
+
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        ~LibGit2Configuration()
+        {
+            Dispose(false);
+        }
+
+        private static unsafe git_config* CreateSnapshot(git_config* config)

```

Comment:
Do we need protection from calling this multiple times? It appears we do not keep a list or stack of snapshots.

Reply:
This method is only used from the constructor, we always create one snapshot and store it in the object as a field (and clean it up after).

To avoid this from being introduced, I'll just inline the `git_config_snapshot` call.